stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/listener/ServerCallHandler.java

```diff
@@ -166,41 +164,49 @@ private boolean isEmptyResponse() {
     }
 
-    void onErrorInvoke(Resource resource, StreamObserver responseObserver, Message error) {
-        if (resource == null) {
+    void onErrorInvoke(ServiceResource resource, StreamObserver responseObserver, Message error) {
+        if (resource == null || resource.getParamDetailList() == null) {
             String message = "Error in listener service definition. onError resource does not exists";
             LOG.error(message);
             throw new ServerRuntimeException(message);
         }
-        List<ParamDetail> paramDetails = resource.getParamDetails();
+        List<ParamDetail> paramDetails = resource.getParamDetailList();
         BValue[] signatureParams = new BValue[paramDetails.size()];
-        signatureParams[0] = getConnectionParameter(resource, responseObserver);
+        signatureParams[0] = getConnectionParameter(resource.getProgramFile(), responseObserver);
         BType errorType = paramDetails.get(1).getVarType();
         BError errorStruct = MessageUtils.getConnectorError((BErrorType) errorType, error.getError());
         signatureParams[1] = errorStruct;
-        BMap<String, BValue> headerStruct = getHeaderStruct(resource);
-        if (headerStruct != null) {
+        BMap<String, BValue> headerStruct = null;
+        if (resource.isHeaderRequired()) {
+            headerStruct = getHeaderStruct(resource.getProgramFile());
             headerStruct.addNativeData(MESSAGE_HEADERS, error.getHeaders());
         }
 
         if (headerStruct != null && signatureParams.length == 3) {
             signatureParams[2] = headerStruct;
         }
         CallableUnitCallback callback = new GrpcCallableUnitCallBack(null);
-        Executor.submit(resource, callback, null, null, signatureParams);
+        Executor.submit(resource.getResource(), callback, null, null, signatureParams);
     }
 
-    void onMessageInvoke(Resource resource, Message request, StreamObserver responseObserver) {
+    void onMessageInvoke(ServiceResource resource, Message request, StreamObserver responseObserver) {
         CallableUnitCallback callback = new GrpcCallableUnitCallBack(responseObserver, isEmptyResponse());
-        Executor.submit(resource, callback, null, null, computeMessageParams(resource, request, responseObserver));
+        Executor.submit(resource.getResource(), callback, null, null, computeMessageParams(resource, request,
+                responseObserver));
     }
 
-    BValue[] computeMessageParams(Resource resource, Message request, StreamObserver responseObserver) {
-        List<ParamDetail> paramDetails = resource.getParamDetails();
+    BValue[] computeMessageParams(ServiceResource resource, Message request, StreamObserver responseObserver) {
+        if (resource == null || resource.getParamDetailList() == null) {
+            String message = "Error when dispatching request. Incorrect service resource definition";
+            LOG.error(message);
+            throw new ServerRuntimeException(message);
```

Comment:
If we are throwing an error with a massage then do we need to log it here? Won't that capture through thrown exception in the upper layer?

Reply:
Yes, we don't need to log it again. I removed it in commit ` [877a9b2](/ballerina-platform/ballerina-lang/pull/13345/commits/877a9b21ad8cc6423910c3377004e286f8fbe8f3)`